File: backend/apps/catalog/interfaces/i_catalog_service.py
"""
Root ABC for the catalog module.

Responsibility (SRP): declare the complete catalog operation contract.
    No HTTP logic, no ORM queries, no serialization — only method signatures.
Depends on: abc.ABC — nothing from the domain or Django.
Pattern: DIP anchor — CatalogService (Singleton) will implement this in S11.
SOLID: DIP · OCP · LSP

Sprint coverage:
    S11 → this file (contract only)
    S11 → CatalogService(ICatalogService, ICatalogClientView, ICatalogAdminView) — Singleton
    S18 → tests mock ICatalogService; views receive the mock without modification (LSP proof)

OCP extension path:
    New catalog operation (e.g. archive_service, duplicate_service) = new @abstractmethod here
    + implementation in CatalogService. ICatalogClientView and ICatalogAdminView remain frozen.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class ICatalogService(ABC):
    """Abstract contract for all catalog use cases."""

    @abstractmethod
    def get_active_services(self, filters: dict | None = None) -> list:
        """
        Return all services with activo=True.
        Optional filters: {'categoria': str, 'busqueda': str}
        Returns: list of ServiceSummary dicts.
        Raises: nothing — returns empty list if no services match.
        """
        ...

    @abstractmethod
    def get_service_detail(self, service_id: int) -> dict:
        """
        Return full detail of a single service.
        Returns: ServiceDetail dict.
        Raises: ServiceNotFound if service_id does not exist or is inactive.
        """
        ...

    @abstractmethod
    def create_service(self, data: dict) -> dict:
        """
        Create a new service entry in the catalog (admin only — enforced at view level).
        Args: data = validated dict from ServiceCreateSerializer
              (nombre: str, descripcion: str, categoria: str).
        Returns: created ServiceDetail dict.
        Raises: ValidationError if nombre already exists.
        """
        ...

    @abstractmethod
    def edit_service(self, service_id: int, data: dict) -> dict:
        """
        Partially update an existing service.
        Args: data = validated dict from ServiceEditSerializer (all fields optional).
        Returns: updated ServiceDetail dict.
        Raises: ServiceNotFound, ValidationError.
        """
        ...

    @abstractmethod
    def toggle_active(self, service_id: int) -> dict:
        """
        Flip the activo field: True → False or False → True.
        Returns: updated ServiceDetail dict with new activo value.
        Raises: ServiceNotFound.
        """
        ...

File: frontend/src/modules/catalog/interfaces/ICatalogService.ts
/**
 * Root contract for all catalog operations in the frontend.
 *
 * Responsibility (SRP): declare the complete catalog operation contract.
 *     No HTTP calls, no state management — only method signatures and shared types.
 * Depends on: nothing — root abstraction.
 * Pattern: DIP anchor — CatalogService (Singleton) will implement this in S11.
 * SOLID: DIP · OCP · LSP
 *
 * OCP extension: new catalog operation = new method signature here + implementation
 *     in CatalogService. ICatalogClientView and ICatalogAdminView remain frozen.
 */

// ─── Shared data shapes ──────────────────────────────────────────────────────

export interface ServiceSummary {
  id: string
  nombre: string
  descripcion: string
  categoria: string
  activo: boolean
  imagenUrl: string
}

export interface ServiceDetail extends ServiceSummary {
  creadoEn: string       // ISO 8601
  actualizadoEn: string  // ISO 8601
}

export interface ServiceFilterOptions {
  categoria?: string
  busqueda?: string
}

export interface ServiceCreatePayload {
  nombre: string
  descripcion: string
  categoria: string
}

export interface ServiceEditPayload {
  nombre?: string
  descripcion?: string
  categoria?: string
}

// ─── Service contract ─────────────────────────────────────────────────────────

export interface ICatalogService {
  /** Return all active services, optionally filtered by category or free-text. */
  getActiveServices(filters?: ServiceFilterOptions): Promise<ServiceSummary[]>

  /** Return full detail of one active service. Throws ServiceNotFound if missing. */
  getServiceDetail(id: string): Promise<ServiceDetail>

  /** Create a new service (admin only — enforced at API level). */
  createService(data: ServiceCreatePayload): Promise<ServiceDetail>

  /** Partially update an existing service. */
  editService(id: string, data: ServiceEditPayload): Promise<ServiceDetail>

  /** Flip the active/inactive state of a service. */
  toggleActive(id: string): Promise<ServiceDetail>
}
